tests/e2e/failover_test.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !cluster_proxy
+
+package e2e
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	_ "google.golang.org/grpc/health"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+)
+
+const (
+	// in sync with how kubernetes uses etcd
+	// https://github.com/kubernetes/kubernetes/blob/release-1.28/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go#L59-L71
+	keepaliveTime    = 30 * time.Second
+	keepaliveTimeout = 10 * time.Second
+	dialTimeout      = 20 * time.Second
+
+	clientRuntime  = 10 * time.Second
+	requestTimeout = 100 * time.Millisecond
+
+	// common assertion
+	expectedMinTotalRequestsCount = 300
+	// happy case assertion
+	expectedMaxFailedRequestsCount = 5
+	// negative case assertion
+	expectedMinFailedRequestsCount = 90
+)
+
+func TestFailoverOnDefrag(t *testing.T) {
+	tcs := []struct {
+		name           string
+		clusterOptions []e2e.EPClusterOption
+
+		disableClientHealthCheck bool
+		assertion                func(t *testing.T, requestVolume, successfulRequestCount int)
```

Comment:
Please avoid a dedicated functions for assertions, instead have two variables, `expectedMinTotalRequestsCount`, `expectedMaxFailedRequestsCount` set on tcs and inline the assetions into TestFailoverOnDefrag function

```
	if tc.expectedMinTotalRequestsCount {
		require.GreaterOrEqual(t, requestVolume, tc.expectedMinTotalRequestsCount)
	}
	if tc.expectedMaxFailedRequestsCount {
		failedRequestCount := requestVolume - successfulRequestCount
		require.LessOrEqual(t, failedRequestCount, expectedMaxFailedRequestsCount)
	}
```

Reply:
Done!